esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+	FlushInterval time.Duration // The flush threshold as duration. Defaults to 10sec.
+
+	Client      *elasticsearch.Client   // The Elasticsearch client.
+	Decoder     BulkResponseJSONDecoder // A custom JSON decoder.
+	DebugLogger BulkIndexerDebugLogger  // An optional logger for debugging.
+
+	// Parameters of the Bulk API.
+	Index               string
+	ErrorTrace          bool
+	FilterPath          []string
+	Header              http.Header
+	Human               bool
+	Pipeline            string
+	Pretty              bool
+	Refresh             string
+	Routing             string
+	Source              []string
+	SourceExcludes      []string
+	SourceIncludes      []string
+	Timeout             time.Duration
+	WaitForActiveShards string
+}
+
+// BulkIndexerStats represents the indexer statistics.
+//
+type BulkIndexerStats struct {
+	NumAdded    uint
+	NumFlushed  uint
+	NumFailed   uint
+	NumIndexed  uint
+	NumCreated  uint
+	NumUpdated  uint
+	NumDeleted  uint
+	NumRequests uint
+}
+
+// BulkIndexerItem represents an indexer item.
+//
+type BulkIndexerItem struct {
+	Index           string
+	Action          string
+	DocumentID      string
+	Body            io.Reader
+	RetryOnConflict *int
+
+	Metadata interface{} // To use eg. in OnSuccess/OnFailure callbacks
+
+	OnSuccess func(item BulkIndexerItem, res BulkIndexerResponseItem)            // Per item
+	OnFailure func(item BulkIndexerItem, res BulkIndexerResponseItem, err error) // Per item
+}
+
+// BulkIndexerResponse represents the Elasticsearch response.
+//
+type BulkIndexerResponse struct {
+	Took      int                                  `json:"took"`
+	HasErrors bool                                 `json:"errors"`
+	Items     []map[string]BulkIndexerResponseItem `json:"items,omitempty"`
+}
+
+// BulkIndexerResponseItem represents the Elasticsearch response item.
+//
+type BulkIndexerResponseItem struct {
+	Index      string `json:"_index"`
+	DocumentID string `json:"_id"`
+	Version    int64  `json:"_version,omitempty"`
+	Result     string `json:"result,omitempty"`
+	Status     int    `json:"status,omitempty"`
+	SeqNo      int64  `json:"_seq_no,omitempty"`
+	PrimTerm   int64  `json:"_primary_term,omitempty"`
+
+	Shards struct {
+		Total      int `json:"total,omitempty"`
+		Successful int `json:"successful,omitempty"`
+		Failed     int `json:"failed,omitempty"`
+	} `json:"_shards,omitempty"`
+
+	Error struct {
+		Type   string `json:"type"`
+		Reason string `json:"reason"`
+		Cause  struct {
+			Type   string `json:"type"`
+			Reason string `json:"reason"`
+		} `json:"caused_by"`
+	} `json:"error,omitempty"`
+}
+
+// BulkResponseJSONDecoder defines the interface for custom JSON decoders.
+//
+type BulkResponseJSONDecoder interface {
+	UnmarshalFromReader(io.Reader, *BulkIndexerResponse) error
+}
+
+// BulkIndexerDebugLogger defines the interface for a debugging logger.
+//
+type BulkIndexerDebugLogger interface {
+	Printf(string, ...interface{})
+}
```

Comment:
We have this problem with our Kafka library. What is debug and what is info we want to show (some state change due to sniffing for example). Just seeing Printf makes me wonder if every log is really 'debug'.

As we have our own interface, would it make sense to have an interface (this one or an additional one) that gives away a little more information for the purpose of structured logging? E.g. if there was an IO error, which host?

Reply:
That's a good point. This "logger", however, is really something only relevant for debugging — effectively, it just wrapped the `fmt.Printf` statements I have sprinkled over the codebase during development.

As there's no common logging interface in Go, it would be rather hard to support something like a structured logging — I've struggled with that couple of times already.

I'm thinking that instead of having the component doing the logging, the _calling code_ should do the logging itself, by using some kind of hooks. This is easy for events like "error indexing item", by using the `OnFailure()` callback. But for more low-level events, like "worker 001 started", we would need to provide specific hooks for them. I'm not sure if low-level events like these are terribly useful for a production system, though?

Note that the component exposes the `BulkIndexerStats`, which is something I imagine a logging/monitoring system would consume...

Reply:
Yeah, I'm hinting at introducing hooks. At least for very important events. But I guess the more important events (e.g. sniffing failed) will be comming from the esclient lib itself.

Reply:
Yes, the more important events will happen on the client level. I'm thinking about removing the `DebugLogger` altogether, to prevent the public API pollution. 

It would be indeed a nice feature to add hooks, which the user can implement and pass to configuration — I'm just not sure I want to cram all of that into a first pull request...

Reply:
In case it's only debug, I'd still prefer to have a Printf style hook over no hook. At least it gives me some logs I can have a look at in case things go wrong.